platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/KotlinDslContainerElementFactoryIntegrationTest.kt

```diff
@@ -0,0 +1,108 @@
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.integration
+
+import org.gradle.kotlin.dsl.accessors.DCL_ENABLED_PROPERTY_NAME
+import org.gradle.kotlin.dsl.fixtures.AbstractKotlinIntegrationTest
+import org.junit.Test
+
+
+class KotlinDslContainerElementFactoryIntegrationTest : AbstractKotlinIntegrationTest() {
+    @Test
+    fun `a named domain object container gets a synthetic element factory function`() {
+        withBuildScript(
+            """
+            plugins {
+                `java`
+            }
+
+            configurations {
+                configuration("myConfiguration") { }
+            }
+
+            sourceSets {
+                sourceSet("mySourceSet") {
+                    println("created my source set!")
+                }
+            }
+        """
+        )
+
+        buildAndFail("dependencies") // no DCL support by default
+            .assertHasErrorOutput("Unresolved reference: configuration")
+
+        with(build("dependencies", enableDclCliFlag)) {
+            assertTasksExecuted(":dependencies")
+            assertOutputContains("myConfiguration\nNo dependencies")
+        }
+
+        with(build("compileMySourceSetJava", enableDclCliFlag)) {
+            assertTasksSkipped(":compileMySourceSetJava")
+            assertOutputContains("created my source set!")
+        }
+    }
+
+    @Test
+    fun `can use custom names for container element factories`() {
+        withKotlinBuildSrc()
+        withFile("buildSrc/src/main/kotlin/myPlugin.gradle.kts", """
+            import org.gradle.declarative.dsl.model.annotations.ElementFactoryName
+
+            abstract class MyExtension {
+                abstract val myElements: NamedDomainObjectContainer<MyElement>
+            }
+
+            @ElementFactoryName("customName")
+            abstract class MyElement(val elementName: String) : Named {
+                override fun getName() = elementName
+            }
+
+            val myExtension = project.extensions.create("myExtension", MyExtension::class.java)
+
+            tasks.register("printNames") {
+                val names = myExtension.myElements.names
+                doFirst {
+                    println(names)
+                }
+            }
+        """.trimIndent())
+
+        withBuildScript("""
+            plugins {
+                id("myPlugin")
+            }
+
+            myExtension {
+                myElements {
+                    customName("one") { }
+                    customName("two") { }
+                }
+            }
+        """.trimIndent())
+
+        buildAndFail("printNames") // no DCL support by default
+            .assertHasErrorOutput("Unresolved reference: customName")
+
+        with(build("printNames", enableDclCliFlag)) {
+            assertTaskExecuted(":printNames")
+            assertOutputContains("[one, two]")
+        }
+    }
+
+    private val enableDclCliFlag =
+        "-D${DCL_ENABLED_PROPERTY_NAME}=true"
```

Comment:
💅 It would be good to also have a test that enables that flag from `gradle.properties`